steel/src/contract.rs

```diff
@@ -0,0 +1,441 @@
+/// [ViewCallInput::into_env].
+///
+/// # Examples
+///
+/// ```no_run
+/// use risc0_steel::{ethereum::EthViewCallEnv, Contract};
+/// use alloy_primitives::{address};
+/// use alloy_sol_types::sol;
+///
+/// # fn main() -> anyhow::Result<()> {
+/// let contract_address = address!("dAC17F958D2ee523a2206206994597C13D831ec7");
+/// sol! {
+/// #[derive(Debug, PartialEq, Eq)]
+/// interface IERC20 {
+///    function balanceOf(address account) external view returns (uint);
+/// }
+/// }
+///
+/// let get_balance = IERC20::balanceOfCall {
+///     account: address!("F977814e90dA44bFA03b6295A0616a897441aceC"),
+/// };
+///
+/// // Host:
+///
+/// let mut env = EthViewCallEnv::from_rpc("https://ethereum-rpc.publicnode.com", None)?;
+/// let mut contract = Contract::preflight(contract_address, &mut env);
+/// contract.call_builder(&get_balance).call()?;
+///
+/// let view_call_input = env.into_zkvm_input()?;
+///
+/// // Guest
+///
+/// let view_call_env = view_call_input.into_env();
+/// let contract = Contract::new(contract_address, &view_call_env);
+/// contract.call_builder(&get_balance).call();
+///
+/// # Ok(())
+/// # }
+/// ```
+///
+/// [EthViewCallInput::into_env]: ethereum::EthViewCallInput::into_env
+/// [EthViewCallEnv::from_rpc]: ethereum::EthViewCallEnv::from_rpc
+pub struct Contract<E> {
+    address: Address,
+    env: E,
+}
+
+impl<'a, H> Contract<&'a GuestViewCallEnv<H>> {
+    pub fn new(address: Address, env: &'a GuestViewCallEnv<H>) -> Self {
+        Self { address, env }
+    }
+
+    /// Initialize a call builder to execute a call on the contract. For more information on usage,
+    /// see [Contract].
+    pub fn call_builder<C: SolCall>(&self, call: &C) -> ViewCallBuilder<&GuestViewCallEnv<H>, C> {
+        ViewCallBuilder::new_sol(self.env, self.address, call)
+    }
+}
+
+#[cfg(feature = "host")]
+impl<'a, P, H> Contract<&'a mut HostViewCallEnv<P, H>>
+where
+    P: Provider,
+{
+    /// Constructor to initialize a [ViewCallEnv] outside of the guest program. When calling
+    /// functions on the contract, the data needed will be fetched through the [Provider], and
+    /// a storage proof of any elements accessed will be generated in
+    /// [ViewCallEnv::into_zkvm_input].
+    ///
+    /// [Provider]: host::provider::Provider
+    pub fn preflight(address: Address, env: &'a mut HostViewCallEnv<P, H>) -> Self {
+        Self { address, env }
+    }
+
+    /// Initialize a call builder to execute a call on the contract. For more information on usage,
+    /// see [Contract].
+    pub fn call_builder<C: SolCall>(
+        &mut self,
+        call: &C,
+    ) -> ViewCallBuilder<&mut HostViewCallEnv<P, H>, C> {
+        ViewCallBuilder::new_sol(&mut self.env, self.address, call)
+    }
+}
+
+/// A builder for calling an Ethereum contract. Once configured, call with [ViewCallBuilder::call].
+#[derive(Debug, Clone)]
+#[must_use]
+pub struct ViewCallBuilder<E, C> {
```

Comment:
While I agree for the longer term, I was hoping this could just be here for one version (to allow a soft deprecation) and then remove it in the following version, and I'll create the issue to track. Point being I know it's a pre-1.0, but hard breaking changes give devs a lot of overhead, so having a deprecation for a version gives them time to switch, and makes it easier.

Reply:
I just talked about this in the standup.
Since this is a pretty big cleanup, we want to keep the code clean, even if we have to make these breaking changes.
We'll make sure we have examples and documentation to help users make the transition.

Reply:
done. It's much cleaner to still keep the type internally imo, so that's what I did, just removed it from public API. 

Also don't love that non-zero return check (seems like there could theoretically be a use to have a non-return function call), and I could have it in a constructor, but seemed cleanest to just have in the transact function. Noting because the error would show up when the tx is called rather than the builder initiated (I think), so wanted to point that out.

[`7a5523d` (#115)](https://github.com/risc0/risc0-ethereum/pull/115/commits/7a5523dab3177a46a882ff09d9bd2152d5dca697)

Reply:
> Also don't love that non-zero return check (seems like there could theoretically be a use to have a non-return function call)

There could be such a use case, but the problem is also that in `transact` we check that the function has returned, which is actually very helpful in notifying the user if they have accidentally called an EOA (a very common error), and we decode the result, which also fails for functions that do not return. Both of these things would need to be addressed before the check could be removed. Since this seems to be a rather esoteric use case, it definitely feels better to give a compile-time error than a cryptic runtime error.